src/shared/bus-util.c

```diff
@@ -552,8 +553,11 @@ int bus_check_peercred(sd_bus *c) {
         if (r < 0)
                 return r;
 
-        if (ucred.uid != 0 && ucred.uid != geteuid())
-                return -EPERM;
+        if (ucred.uid != 0) {
+                euid = geteuid();
+                if (euid != 0 && ucred.uid != euid)
+                        return -EPERM;
+        }
```

Comment:
hmm? doesn't this mean that anyone can connect to root's --user instance this way? i.e. when this code runs in the --user instance of systemd, and an unpriv client connects, it will see ucred.uid != 0, hence the bigger if block is run. Then euid == 0 hence the inner if block is not run, hence boom, unpriv clients got access to priv systemd --user....

What am I missing?

Reply:
Ah sorry, I was focusing on the code path coming from `bus_connect_user_systemd()`, which is used by the client (systemctl) and I missed the server cases.

Upon further reflection, I noticed this code is really written from the server's point of view (i.e. checking `ucred.uid` to see if the client connecting to it is root or not), so now I'm thinking it's actually OK to just drop the check from the client side (i.e. from `bus_connect_user_systemd()`) and rely on filesystem permissions (and on the server side of the check) instead.

I updated the PR to do that instead. See if you agree with that approach. PTAL.

Reply:
So I actualy deliberately added the code to authentication communication in both directions. I figured it is actually nice if both sides validate that they can trust the peer, so that you cannot trick the client to talk to something else believing it was PID 1. 

And that requirement actually *does* matter: some distros patched things for a while so that XDG_RUNTIME_DIR on su/sudo would be retained, so it would point to the origin user's original value, so that root could think it would talk to root but actually talked to some unpriv user's systemd --user instance.

So I am curious, what's the usecase for what you are trying to do?

I have the suspicion if there's a good reason for what you are trying to do (i.e. as opposed to maybe suggests root to use su/sudo first to transition to the user's context first) then it needs to be made explicit somehow, so that no accidental connections to some other user's manager can't happen. For example, checking some new env var with secure_getenv() that allows turning off the authentication would be OK I guess. `parse_boolean(secure_getenv("SYSTEMD_CONNECT_TO_UNTRUSTED_BUS"))` or so?

Reply:
The original use case was about monitoring resource usage in slices created by users.

But then I'm not really sure this would be the best approach here... Since overriding `XDG_RUNTIME_DIR` would be somewhat useful to get that information from a single user, perhaps a more general solution is to extend systemctl so it could give more visibility into the services/slices managed by potentially all users... (perhaps even in a single invocation.)  Wouldn't that make sense?

Cheers,
Filipe